src/background/host.ts

```diff
@@ -1,26 +1,87 @@
+    #friends = new Map<string, FriendConnected>();
+
+    public personalData: Friend | null = null;
     // eslint-disable-next-line @typescript-eslint/no-empty-function
     public onChangeFriends: (count: number, delta: number) => void = () => {};
 
-    #handleConnect = (conn: Peer.DataConnection): void => {
-        this.#friends.add(conn);
-        this.onChangeFriends(this.#friends.size, 1);
+    #ensureConnections = (friends: Friend[], conn: Peer.DataConnection): void => {
+        if (!this.personalData) {
+            return;
+        }
+
+        for (const member of friends) {
+            const friend = this.#friends.get(member.id);
+
+            if (member.id === this.personalData.id) {
+                // do nothing regarding our own connection
+            } else if (friend) {
+                // update member data without touching `friend.conn`
+                this.#friends.set(member.id, {
+                    ...member,
+                    ...friend
+                });
+            } else if (member.id === conn.peer) {
+                // insert the person we connected to, now that we know their info
+                this.#friends.set(conn.peer, {
+                    ...member,
+                    conn
+                });
+                this.onChangeFriends(this.#friends.size, 1);
+            } else {
+                // for new people, just go connect to them (& hopefully they aren't doing the same)
+                this.connect(member.id);
+            }
+        }
+    };
+
+    #handleConnect = (conn: Peer.DataConnection, isIncoming: boolean): void => {
+        if (!this.personalData) {
+            return;
+        }
+
+        this.#log(`Handling ${isIncoming ? "in" : "out"} connection`, conn.metadata);
+
+        if (isIncoming) {
+            // `conn.metadata` will always be the `Friend` of the **sender**, so when we handle the outgoing request,
+            // we avoid writing ourselves as a friend (and therefore emitting events to ourselves & getting
+            // infinite-looped).
+            const friend: FriendConnected = {
+                ...(conn.metadata as Friend),
+                conn
+            };
+
+            this.#friends.set(conn.peer, friend);
+            this.onChangeFriends(this.#friends.size, 1);
+        }
+
+        conn.on("data", (data: RemoteMessage) => {
+            if (!this.personalData) {
+                return;
+            }
+            // We need to randomize `#ensureConnections()` so that two peers don't try to connect to each other at
+            // the exact same time (immediately upon receiving the message).
+            const timer = Math.round(Math.random() * 1000);
+            this.#log(`Received data from ${conn.peer} (will wait ${timer}ms for connections)`, data);
 
-        conn.on("data", (data: Message) => {
             this.#port.postMessage(data);
+            setTimeout(() => this.#ensureConnections(data.friends, conn), timer);
```

Comment:
We basically want to keep everyone in sync all the time .. right now. Until we build a muting/host-control-only feature.

I think being out-of-sync is a pretty fatal problem, and we should course-correct as much as possible until everyone agrees. The way I have it set up ... is that any change to state will tell everyone the state as it should be. Peers who already follow that state (with 250ms leeway on `video.currentTime`) do nothing .... so it's only peers who we deem too-far-from-the-desired-state that actually do work on one of these messages.

Reply:
It seems like we might not be on the same page? If we're planning on ever supporting people being in different states, then this design will have to change to look like what I already have. I also think my design is more efficient even though it's admittedly more complex

Reply:
I don't necessarily think we would be prevented from supporting that if we still emit video events to sync the peer list.

One thing I could imagine seeing in the UI is the play time of each person. That way, you can see if anyone has de-sync'd themself from the stream ... and we could support that pretty easily if it is the recipient who ignores a `muted` person, rather than the sender.